ksql-common/src/main/java/io/confluent/ksql/metrics/StreamsErrorCollector.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.metrics;
+
+import static io.confluent.ksql.metrics.MetricCollectors.CONSUMER_FAILED_MESSAGES;
+import static io.confluent.ksql.metrics.MetricCollectors.CONSUMER_FAILED_MESSAGES_PER_SEC;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import io.confluent.common.utils.Time;
+import io.confluent.ksql.metrics.TopicSensors.SensorMetric;
+import io.confluent.ksql.metrics.TopicSensors.Stat;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.MeasurableStat;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.Total;
+
+public final class StreamsErrorCollector implements MetricCollector {
+  private static final Map<String, StreamsErrorCollector> COLLECTORS = Maps.newConcurrentMap();
```

Comment:
I can't think of a great way to get rid of the map altogether. We need a way to map from the app id to the collector from the error handler. One option would be to put it in the configs, since those get passed around in the ProcessorContext. That felt brittle since I'm not sure the config was intended to be used that way.

We could move this state out to a separate singleton, if that's what you're suggesting. Not sure where dependency injection comes in though.

Reply:
I wasn't suggesting moving to another singleton. I was just wondering if it was possible to store this state on an object, and pass the object around to where its needed, (i.e. inject the dependency).

If it's not possible, then we'll go with this :D